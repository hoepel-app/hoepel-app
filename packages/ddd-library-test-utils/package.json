{
  "name": "@hoepel.app/ddd-library-test-utils",
  "version": "0.3.113",
  "main": "dist/index.js",
  "license": "GPL-3.0-or-later",
  "devDependencies": {
    "typescript": "^4.0.5"
  },
  "scripts": {
    "build": "tsc"
  },
  "private": true
}
