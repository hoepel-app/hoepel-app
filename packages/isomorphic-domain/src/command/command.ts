export type CommandRejected = { status: 'rejected' }
export type CommandAccepted = { status: 'accepted' }
export type CommandResult = CommandAccepted | CommandRejected

export type Command<T> = {
  toProps(): T
}
