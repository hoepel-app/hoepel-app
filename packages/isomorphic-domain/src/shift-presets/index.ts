export { ShiftPreset, ShiftPresetProps } from './shift-preset'
export { ShiftPresets, ShiftPresetsProps } from './shift-presets'
export { ShiftPresetsRepository } from './shift-presets.repository'
export { AddShiftPresetCommand } from './commands/add-shift-preset.command'
export { RemoveShiftPresetCommand } from './commands/remove-shift-preset.command'
export { RenameShiftPresetCommand } from './commands/rename-shift-preset.command'
export { ChangeShiftPresetPriceCommand } from './commands/change-shift-preset-price.command'
export { ChangeShiftPresetDescriptionCommand } from './commands/change-shift-preset-description.command'
export { ChangeShiftPresetLocationCommand } from './commands/change-shift-preset-location.command'
export { ChangeChildrenCanAttendShiftPresetCommand } from './commands/change-children-can-attend-shift-preset.command'
export { ChangeCrewMembersCanAttendShiftPresetCommand } from './commands/change-crew-members-can-attend-shift-preset.command'
export { ChangeShiftPresetStartAndEndCommand } from './commands/change-shift-preset-start-and-end.command'
export { ShiftPresetsApplicationService } from './shift-presets.application-service'
