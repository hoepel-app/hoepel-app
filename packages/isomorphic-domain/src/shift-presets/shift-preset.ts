import { Price, StartAndEndTime } from '@hoepel.app/types'

export type ShiftPresetProps = {
  readonly name: string
  readonly priceCents: number
  readonly childrenCanAttend: boolean
  readonly crewMembersCanAttend: boolean

  /** Can be empty string */
  readonly location: string

  /** Can be empty string */
  readonly description: string

  /** Start time of shifts created using this shift preset, in minutes since midnight */
  readonly startMinutesSinceMidnight: number

  /** End time of shifts created using this shift preset, in minutes since midnight */
  readonly endMinutesSinceMidnight: number
}

export class ShiftPreset {
  private constructor(private readonly props: ShiftPresetProps) {}

  get price(): Price {
    return Price.fromCents(this.props.priceCents)
  }

  get name(): string {
    return this.props.name
  }

  get childrenCanAttend(): boolean {
    return this.props.childrenCanAttend
  }

  get crewMembersCanAttend(): boolean {
    return this.props.crewMembersCanAttend
  }

  get location(): string {
    return this.props.location
  }

  get description(): string {
    return this.props.description
  }

  get startAndEndTime(): StartAndEndTime {
    // TODO extract this logic to LocalTime/StartAndEndTime
    const startMinutes = this.props.startMinutesSinceMidnight % 60
    const startHours =
      (this.props.startMinutesSinceMidnight - startMinutes) / 60
    const endMinutes = this.props.endMinutesSinceMidnight % 60
    const endHours = (this.props.endMinutesSinceMidnight - endMinutes) / 60

    return new StartAndEndTime({
      start: { hour: startHours, minute: startMinutes },
      end: { hour: endHours, minute: endMinutes },
    })
  }

  static createEmpty(name: string): ShiftPreset {
    return new ShiftPreset({
      priceCents: 0,
      name,
      childrenCanAttend: true,
      crewMembersCanAttend: true,
      description: '',
      location: '',
      startMinutesSinceMidnight: 540, // 9:00
      endMinutesSinceMidnight: 1020, // 17:00
    })
  }

  static fromProps(props: ShiftPresetProps): ShiftPreset {
    return new ShiftPreset(props)
  }

  toProps(): ShiftPresetProps {
    return this.props
  }

  withName(name: string): ShiftPreset {
    return ShiftPreset.fromProps({
      ...this.toProps(),
      name,
    })
  }

  withPrice(price: Price): ShiftPreset {
    return ShiftPreset.fromProps({
      ...this.toProps(),
      priceCents: price.totalCents,
    })
  }

  withChildrenCanAttend(childrenCanAttend: boolean): ShiftPreset {
    return ShiftPreset.fromProps({
      ...this.toProps(),
      childrenCanAttend,
    })
  }

  withCrewMembersCanAttend(crewMembersCanAttend: boolean): ShiftPreset {
    return ShiftPreset.fromProps({
      ...this.toProps(),
      crewMembersCanAttend,
    })
  }

  withLocation(location: string): ShiftPreset {
    return ShiftPreset.fromProps({
      ...this.toProps(),
      location,
    })
  }

  withDescription(description: string): ShiftPreset {
    return ShiftPreset.fromProps({
      ...this.toProps(),
      description,
    })
  }

  withStartAndEndTime(startAndEnd: StartAndEndTime): ShiftPreset {
    const startMinutesSinceMidnight =
      startAndEnd.start.minute + startAndEnd.start.hour * 60
    const endMinutesSinceMidnight =
      startAndEnd.end.minute + startAndEnd.end.hour * 60

    return ShiftPreset.fromProps({
      ...this.toProps(),
      startMinutesSinceMidnight,
      endMinutesSinceMidnight,
    })
  }
}
