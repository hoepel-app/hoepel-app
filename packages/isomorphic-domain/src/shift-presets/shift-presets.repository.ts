import { Observable } from 'rxjs'
import { ShiftPresets } from './shift-presets'

export type ShiftPresetsRepository = {
  getForTenant(tenantId: string): Observable<ShiftPresets>
  put(entity: ShiftPresets): Promise<void>
}
