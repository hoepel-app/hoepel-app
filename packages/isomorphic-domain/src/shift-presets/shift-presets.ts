import { Aggregate } from '@hoepel.app/ddd-library'
import { ShiftPresetProps, ShiftPreset } from './shift-preset'
import { Price } from '@hoepel.app/types'

export type ShiftPresetsProps = {
  readonly shiftPresets: readonly ShiftPresetProps[]
  readonly tenantId: string
}

export class ShiftPresets implements Aggregate {
  private constructor(private readonly props: ShiftPresetsProps) {}

  get id(): string {
    return `${this.tenantId}-shift-presets`
  }

  get tenantId(): string {
    return this.props.tenantId
  }

  get presets(): readonly ShiftPreset[] {
    return this.props.shiftPresets
      .map((props) => ShiftPreset.fromProps(props))
      .sort((a, b) => a.name.localeCompare(b.name))
  }

  get names(): ReadonlySet<string> {
    return new Set(this.presets.map((group) => group.name))
  }

  get namesSorted(): readonly string[] {
    return [...this.names].sort()
  }

  get numPresets(): number {
    return this.props.shiftPresets.length
  }

  findPresetWithName(name: string): ShiftPreset | null {
    return this.presets.find((preset) => preset.name === name) || null
  }

  hasPresetWithName(name: string): boolean {
    return this.findPresetWithName(name) !== null
  }

  mayAddPreset(preset: ShiftPreset): boolean {
    return !this.hasPresetWithName(preset.name)
  }

  toProps(): ShiftPresetsProps {
    return this.props
  }

  static fromProps(props: ShiftPresetsProps): ShiftPresets {
    return new ShiftPresets(props)
  }

  static createEmpty(tenantId: string): ShiftPresets {
    return new ShiftPresets({
      tenantId,
      shiftPresets: [],
    })
  }

  withPresetAdded(preset: ShiftPreset): ShiftPresets {
    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((p) => p.name !== preset.name)
          .map((p) => p.toProps()),
        preset.toProps(),
      ],
    })
  }

  withPresetRemoved(presetName: string): ShiftPresets {
    if (!this.hasPresetWithName(presetName)) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== presetName)
          .map((preset) => preset.toProps()),
      ],
    })
  }

  withPresetRenamed(oldName: string, newName: string): ShiftPresets {
    const newPreset = this.findPresetWithName(oldName)?.withName(newName)

    if (newPreset == null || this.hasPresetWithName(newName)) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== oldName)
          .map((preset) => preset.toProps()),
        newPreset.toProps(),
      ],
    })
  }

  withPresetPriceChanged(presetName: string, newPrice: Price): ShiftPresets {
    const newPreset = this.findPresetWithName(presetName)?.withPrice(newPrice)

    if (newPreset == null) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== presetName)
          .map((preset) => preset.toProps()),
        newPreset.toProps(),
      ],
    })
  }

  withPresetLocationChanged(
    presetName: string,
    newLocation: string
  ): ShiftPresets {
    const newPreset = this.findPresetWithName(presetName)?.withLocation(
      newLocation
    )

    if (newPreset == null) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== presetName)
          .map((preset) => preset.toProps()),
        newPreset.toProps(),
      ],
    })
  }

  withPresetDescriptionChanged(
    presetName: string,
    newDescription: string
  ): ShiftPresets {
    const newPreset = this.findPresetWithName(presetName)?.withDescription(
      newDescription
    )

    if (newPreset == null) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== presetName)
          .map((preset) => preset.toProps()),
        newPreset.toProps(),
      ],
    })
  }

  withPresetCrewMembersCanAttendChanged(
    presetName: string,
    crewMembersCanAttend: boolean
  ): ShiftPresets {
    const newPreset = this.findPresetWithName(
      presetName
    )?.withCrewMembersCanAttend(crewMembersCanAttend)

    if (newPreset == null) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== presetName)
          .map((preset) => preset.toProps()),
        newPreset.toProps(),
      ],
    })
  }

  withPresetChildrenCanAttendChanged(
    presetName: string,
    childrenCanAttend: boolean
  ): ShiftPresets {
    const newPreset = this.findPresetWithName(
      presetName
    )?.withChildrenCanAttend(childrenCanAttend)

    if (newPreset == null) {
      return this
    }

    return ShiftPresets.fromProps({
      ...this.toProps(),
      shiftPresets: [
        ...this.presets
          .filter((preset) => preset.name !== presetName)
          .map((preset) => preset.toProps()),
        newPreset.toProps(),
      ],
    })
  }
}
