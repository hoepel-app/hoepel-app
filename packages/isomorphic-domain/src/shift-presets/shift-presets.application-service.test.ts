import { CommandMetadata } from '@hoepel.app/ddd-library'
import { of } from 'rxjs'
import { first } from 'rxjs/operators'
import { ShiftPresets } from './shift-presets'
import { ShiftPreset } from './shift-preset'
import { Price } from '@hoepel.app/types'
import { ShiftPresetsApplicationService } from './shift-presets.application-service'
import { AddShiftPresetCommand } from './commands/add-shift-preset.command'
import { RemoveShiftPresetCommand } from './commands/remove-shift-preset.command'
import { RenameShiftPresetCommand } from './commands/rename-shift-preset.command'
import { ChangeShiftPresetPriceCommand } from './commands/change-shift-preset-price.command'

import '@hoepel.app/ddd-library-test-utils'
import { ChangeShiftPresetLocationCommand } from './commands/change-shift-preset-location.command'
import { ChangeShiftPresetDescriptionCommand } from './commands/change-shift-preset-description.command'
import { ChangeCrewMembersCanAttendShiftPresetCommand } from './commands/change-crew-members-can-attend-shift-preset.command'
import { ChangeChildrenCanAttendShiftPresetCommand } from './commands/change-children-can-attend-shift-preset.command'

describe('ShiftsPresetsApplicationService', () => {
  const examplePresets = (tenantId: string): ShiftPresets =>
    ShiftPresets.createEmpty(tenantId)
      .withPresetAdded(ShiftPreset.createEmpty('Afternoon'))
      .withPresetAdded(
        ShiftPreset.createEmpty('External activity').withPrice(
          Price.fromCents(3000)
        )
      )
      .withPresetAdded(
        ShiftPreset.createEmpty('Crew activity').withChildrenCanAttend(false)
      )

  const commandMetadata: CommandMetadata = {
    commandId: 'my-command-id-123',
    requestedBy: {
      email: 'test@example.org',
      type: 'user',
      uid: 'my-uid-123',
    },
    tenantId: 'some-tenant-id',
    timestamp: new Date('2020-04-25T19:43:29.161Z'),
  }

  describe('findShiftPresets', () => {
    it('finds shift presets for a tenant', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)

      const found = await service
        .findShiftPresets('my-tenant-id-here')
        .pipe(first())
        .toPromise()

      expect(repo.getForTenant).toHaveBeenCalledTimes(1)
      expect(repo.getForTenant).toHaveBeenLastCalledWith('my-tenant-id-here')
      expect(found).toEqual(examplePresets('my-tenant-id-here'))
    })
  })

  describe('addShiftPreset', () => {
    it('rejects when adding duplicate shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const newPreset = ShiftPreset.createEmpty('External activity')
      const command = AddShiftPresetCommand.create(newPreset, commandMetadata)

      const commandResult = await service.addShiftPreset(command)

      expect(commandResult).toBeRejected()
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('adds a new shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const newPreset = ShiftPreset.createEmpty('Noon')
      const command = AddShiftPresetCommand.create(newPreset, commandMetadata)

      const commandResult = await service.addShiftPreset(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('removeShiftPreset', () => {
    it('rejects when removing shift preset with unknown name', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = RemoveShiftPresetCommand.create(
        "I don't exist",
        commandMetadata
      )

      const commandResult = await service.removeShiftPreset(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name I don't exist does not exist`
      )
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('removes a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = RemoveShiftPresetCommand.create(
        'Crew activity',
        commandMetadata
      )

      const commandResult = await service.removeShiftPreset(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('renameShiftPreset', () => {
    it('rejects when renaming shift preset with unknown name', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = RenameShiftPresetCommand.create(
        "I don't exist",
        'New Name',
        commandMetadata
      )

      const commandResult = await service.renameShiftPreset(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'I don't exist' does not exist`
      )
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('rejects when renaming shift preset to an existing name', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = RenameShiftPresetCommand.create(
        'Crew activity',
        'External activity',
        commandMetadata
      )

      const commandResult = await service.renameShiftPreset(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'External activity' does already exist`
      )
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('renames a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = RenameShiftPresetCommand.create(
        'Crew activity',
        'My new name for Crew Activity',
        commandMetadata
      )

      const commandResult = await service.renameShiftPreset(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('changeShiftPresetPrice', () => {
    it('rejects when shift preset with name does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeShiftPresetPriceCommand.create(
        'Something here',
        Price.fromCents(6666),
        commandMetadata
      )

      const commandResult = await service.changeShiftPresetPrice(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'Something here' does not exist`
      )
    })

    it('changes the price of a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeShiftPresetPriceCommand.create(
        'Crew activity',
        Price.fromCents(3450),
        commandMetadata
      )

      const commandResult = await service.changeShiftPresetPrice(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('changeShiftPresetLocation', () => {
    it('rejects when shift preset with name does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeShiftPresetLocationCommand.create(
        'Something here',
        'My new location',
        commandMetadata
      )

      const commandResult = await service.changeShiftPresetLocation(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'Something here' does not exist`
      )
    })

    it('changes the location of a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeShiftPresetLocationCommand.create(
        'Crew activity',
        'Bellewaerde',
        commandMetadata
      )

      const commandResult = await service.changeShiftPresetLocation(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('changeShiftPresetDescription', () => {
    it('rejects when shift preset with name does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeShiftPresetDescriptionCommand.create(
        'Something here',
        'My new location',
        commandMetadata
      )

      const commandResult = await service.changeShiftPresetDescription(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'Something here' does not exist`
      )
    })

    it('changes the description of a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeShiftPresetDescriptionCommand.create(
        'External activity',
        'Swimming',
        commandMetadata
      )

      const commandResult = await service.changeShiftPresetDescription(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('changeCrewMembersCanAttendPreset', () => {
    it('rejects when shift preset with name does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeCrewMembersCanAttendShiftPresetCommand.create(
        'Something here',
        false,
        commandMetadata
      )

      const commandResult = await service.changeCrewMembersCanAttendPreset(
        command
      )

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'Something here' does not exist`
      )
    })

    it('changes whether crew members can attend a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeCrewMembersCanAttendShiftPresetCommand.create(
        'Crew activity',
        false,
        commandMetadata
      )

      const commandResult = await service.changeCrewMembersCanAttendPreset(
        command
      )

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('changeChildrenCanAttendPreset', () => {
    it('rejects when shift preset with name does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeChildrenCanAttendShiftPresetCommand.create(
        'Something here',
        false,
        commandMetadata
      )

      const commandResult = await service.changeChildrenCanAttendPreset(command)

      expect(commandResult).toBeRejectedWithReason(
        `A shift preset with the name 'Something here' does not exist`
      )
    })

    it('changes whether children can attend a shift preset', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(examplePresets(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ShiftPresetsApplicationService(repo)
      const command = ChangeChildrenCanAttendShiftPresetCommand.create(
        'Crew activity',
        true,
        commandMetadata
      )

      const commandResult = await service.changeChildrenCanAttendPreset(command)

      expect(commandResult).toBeAccepted()
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })
})
