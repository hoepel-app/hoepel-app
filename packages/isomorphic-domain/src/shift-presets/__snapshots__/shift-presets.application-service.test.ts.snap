// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShiftsPresetsApplicationService addShiftPreset adds a new shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Crew activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Noon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService changeChildrenCanAttendPreset changes whether children can attend a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Crew activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService changeCrewMembersCanAttendPreset changes whether crew members can attend a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": false,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Crew activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService changePresetStartAndEnd changes whether children can attend a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1359,
          "location": "",
          "name": "Crew activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 355,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService changeShiftPresetDescription changes the description of a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Crew activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "Swimming",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService changeShiftPresetLocation changes the location of a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "Bellewaerde",
          "name": "Crew activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService changeShiftPresetPrice changes the price of a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Crew activity",
          "priceCents": 3450,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService removeShiftPreset removes a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ShiftsPresetsApplicationService renameShiftPreset renames a shift preset 1`] = `
Array [
  ShiftPresets {
    "props": Object {
      "shiftPresets": Array [
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "Afternoon",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": true,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "External activity",
          "priceCents": 3000,
          "startMinutesSinceMidnight": 540,
        },
        Object {
          "childrenCanAttend": false,
          "crewMembersCanAttend": true,
          "description": "",
          "endMinutesSinceMidnight": 1020,
          "location": "",
          "name": "My new name for Crew Activity",
          "priceCents": 0,
          "startMinutesSinceMidnight": 540,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;
